arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/transients/StackHeightCheck.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright Consensys Software Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package net.consensys.linea.zktracer.module.hub.transients;
+
+import static com.google.common.base.Preconditions.*;
+import static net.consensys.linea.zktracer.runtime.stack.Stack.MAX_STACK_SIZE;
+
+import lombok.EqualsAndHashCode;
+import net.consensys.linea.zktracer.container.ModuleOperation;
+
+@EqualsAndHashCode(onlyExplicitlyIncluded = true, callSuper = false)
+public class StackHeightCheck extends ModuleOperation {
```

Comment:
I would suggest using an `Integer` as the key to the `StackedSet` since the only field you're using here is an int. If you have lots of these kinds of checks you will create a new key every time which causes lots of memory churn.

`Integer` has a cache within the JVM itself from -128 to 127 but it's also allowed to run further with a flag. Taken from `java.lang.Integer` class:
```
    private static final class IntegerCache {
        static final int low = -128;
        static final int high;

        @Stable
        static final Integer[] cache;
        static Integer[] archivedCache;

        static {
            // high value may be configured by property
            int h = 127;
            String integerCacheHighPropValue =
                VM.getSavedProperty("java.lang.Integer.IntegerCache.high");
            if (integerCacheHighPropValue != null) {
                try {
                    h = Math.max(parseInt(integerCacheHighPropValue), 127);
                    // Maximum array size is Integer.MAX_VALUE
                    h = Math.min(h, Integer.MAX_VALUE - (-low) -1);
                } catch( NumberFormatException nfe) {
                    // If the property cannot be parsed into an int, ignore it.
                }
            }
            high = h;

            // Load IntegerCache.archivedCache from archive, if possible
            CDS.initializeFromArchive(IntegerCache.class);
            int size = (high - low) + 1;

            // Use the archived cache if it exists and is large enough
            if (archivedCache == null || size > archivedCache.length) {
                Integer[] c = new Integer[size];
                int j = low;
                for(int i = 0; i < c.length; i++) {
                    c[i] = new Integer(j++);
                }
                archivedCache = c;
            }
            cache = archivedCache;
            // range [-128, 127] must be interned (JLS7 5.1.7)
            assert IntegerCache.high >= 127;
        }

        private IntegerCache() {}
    }
```

Reply:
The StackHeightCheck StackedSet is a small one, the maximum length is 256. I don't think it is worth using the Integer inner cache here. 
The cache was specifically created to reduce the auto-boxing overhead of Integer to int, and works only with Integer.valueOf().

Reply:
I could be wrong but I thought WcpOperations were being created 2x per opcode execution, hence you would do lots of key lookups during tracing. Exactly! My plan was to compute the comparison `int` and store it in an `Integer` which would be then used as a cached key. It is a permanent cache so that's a downside.
Anyway this is based on what I thought the behaviour was, maybe it's worth to have another round of performance analysis before doing that

Reply:
In reality we could extract the `StackedSet` portion and make it generic and then extend it to a more specialized `StackedSetOfModuleOperations` with more module specific operations. @letypequividelespoubelles @ahamlat @lu-pinto what do you think ?

Reply:
> The StackHeightCheck StackedSet is a small one, the maximum length is 256

That isn't quite true. For overflows the maximum size of this set is `1026`. For  underflows the max size is `(1 + 17) * 1025 = 18450`

Reply:
> n reality we could extract the StackedSet logic and make it generic. Something that has

    a set of committed objects
    a set of as of current objects
    methods to enter, pop, ...

and then extend it to a more specialized StackedSetOfModuleOperations with more module specific operations like lineCount etc ... @letypequividelespoubelles @ahamlat @lu-pinto what do you think ?

Yes, it's the right solution. More we don't need to implement the line counting for the StackedSetOp

Reply:
https://github.com/Consensys/linea-tracer/pull/1242/commits/2caecd04ee5d5016df1f4f59975c6e59048f5f29

Reply:
>  I take it the 256 you reference is the initial allocation. Which tbh we may as well raise to 1024.
May be we should review the naming then, because in the constructor, the two arguments are : `expectedConflationNumberOperations` and `expectedTransactionNumberOperations`. I would suggest to replace expected with initial.

Reply:
>which in turn require as keys things implement the ModuleOperation interface
Alright that makes sense. I completely missed this and just assumed `StackedSet` was generic. Yes I agree that having a class that is generic and another that extends from it that does more specialized checks on `add` is the right way to go.